// @layerzerolabs/protocol-devtools@1.1.4/dist/index.js
'use strict';

var devtools = require('@layerzerolabs/devtools');
var zod = require('zod');
var ioDevtools = require('@layerzerolabs/io-devtools');

// src/dvn/config.ts
var configureDVN = async (graph, createSdk) => devtools.flattenTransactions([await configureDVNDstConfig(graph, createSdk)]);
var configureDVNDstConfig = async (graph, createSdk) => devtools.flattenTransactions(
  await Promise.all(
    graph.connections.map(async ({ vector: { from, to }, config }) => {
      const sdk = await createSdk(from);
      const dstConfig = await sdk.getDstConfig(to.eid);
      if (devtools.isDeepEqual(dstConfig, config.dstConfig)) {
        return [];
      }
      return [await sdk.setDstConfig(to.eid, config.dstConfig)];
    })
  )
);
var DVNDstConfigSchema = zod.z.object({
  gas: devtools.UIntBigIntSchema,
  multiplierBps: devtools.UIntBigIntSchema,
  floorMarginUSD: devtools.UIntBigIntSchema
});
var createEndpointV2Logger = () => ioDevtools.createModuleLogger("EndpointV2");
var withEndpointV2Logger = ioDevtools.createWithAsyncLogger(createEndpointV2Logger);
var configureEndpointV2RegisterLibraries = withEndpointV2Logger(
  devtools.createConfigureMultiple(
    withEndpointV2Logger(
      async (graph, createSdk) => {
        const logger = createEndpointV2Logger();
        const librariesByEndpoint = graph.connections.reduce(
          (librariesByEndpoint2, { vector: { from }, config }) => librariesByEndpoint2.set(
            from,
            librariesByEndpoint2.getOrElse(from, () => /* @__PURE__ */ new Set()).add(config.defaultReceiveLibrary).add(config.defaultSendLibrary)
          ),
          new devtools.OmniPointMap()
        );
        graph.contracts.forEach(({ point, config }) => {
          var _a;
          (_a = config == null ? void 0 : config.readChannelConfigs) == null ? void 0 : _a.forEach(
            ({ defaultReadLibrary }) => librariesByEndpoint.getOrElse(point, () => /* @__PURE__ */ new Set()).add(defaultReadLibrary)
          );
        });
        const omniTransactions = [];
        logger.verbose(`Checking libraries for registration`);
        for (const [from, libraries] of librariesByEndpoint) {
          const sdk = await createSdk(from);
          const label = devtools.formatOmniPoint(from);
          for (const address of libraries) {
            const isRegistered = await sdk.isRegisteredLibrary(address);
            logger.verbose(`Checking library ${address} for registration on ${label}`);
            if (isRegistered) {
              logger.verbose(`Library ${address} is already registered on ${label}`);
              continue;
            }
            logger.verbose(`Registering library ${address} on ${label}`);
            omniTransactions.push(await sdk.registerLibrary(address));
          }
        }
        return omniTransactions;
      },
      {
        onStart: (logger) => logger.verbose(`Checking register libraries configuration`),
        onSuccess: (logger) => logger.verbose(`${ioDevtools.printBoolean(true)} Checked register libraries configuration`),
        onError: (logger, _graph, error) => logger.error(`Failed to check register libraries configuration: ${error}`)
      }
    )
  )
);
var configureEndpointV2DefaultReceiveLibraries = withEndpointV2Logger(
  devtools.createConfigureEdges(
    withEndpointV2Logger(
      async ({ vector: { from, to }, config }, sdk) => {
        const logger = createEndpointV2Logger();
        const label = devtools.formatOmniVector({ from, to });
        const address = await sdk.getDefaultReceiveLibrary(to.eid);
        logger.verbose(`Checking default receive library for ${label}`);
        if (config.defaultReceiveLibrary === address) {
          logger.verbose(`Default receive library is already set to ${address} for ${label}, skipping`);
          return [];
        }
        logger.info(`Setting default receive library to ${config.defaultReceiveLibrary} for ${label}`);
        return [
          await sdk.setDefaultReceiveLibrary(
            to.eid,
            config.defaultReceiveLibrary,
            config.defaultReceiveLibraryGracePeriod
          )
        ];
      },
      {
        onStart: (logger, [{ vector }]) => logger.verbose(`Checking default receive libraries for ${devtools.formatOmniVector(vector)}`),
        onSuccess: (logger, [{ vector }]) => logger.verbose(
          `${ioDevtools.printBoolean(true)} Checked default receive libraries for ${devtools.formatOmniVector(vector)}`
        ),
        onError: (logger, [{ vector }], error) => logger.error(`Failed to check default receive libraries for ${devtools.formatOmniVector(vector)}: ${error}`)
      }
    )
  )
);
var configureEndpointV2DefaultSendLibraries = withEndpointV2Logger(
  devtools.createConfigureEdges(
    withEndpointV2Logger(
      async ({ vector: { from, to }, config }, sdk) => {
        const logger = createEndpointV2Logger();
        const label = devtools.formatOmniVector({ from, to });
        const address = await sdk.getDefaultSendLibrary(to.eid);
        logger.verbose(`Checking default send library for ${label}`);
        if (config.defaultSendLibrary === address) {
          logger.verbose(`Default send library is already set to ${address} for ${label}, skipping`);
          return [];
        }
        logger.info(`Setting default send library to ${config.defaultSendLibrary} for ${label}`);
        return [await sdk.setDefaultSendLibrary(to.eid, config.defaultSendLibrary)];
      },
      {
        onStart: (logger, [{ vector }]) => logger.verbose(`Checking default send libraries for ${devtools.formatOmniVector(vector)}`),
        onSuccess: (logger, [{ vector }]) => logger.verbose(
          `${ioDevtools.printBoolean(true)} Checked default send libraries for ${devtools.formatOmniVector(vector)}`
        ),
        onError: (logger, [{ vector }], error) => logger.error(`Failed to check default send libraries for ${devtools.formatOmniVector(vector)}: ${error}`)
      }
    )
  )
);
var configureEndpointV2DefaultReadLibraries = withEndpointV2Logger(
  devtools.createConfigureNodes(
    withEndpointV2Logger(
      async ({ config, point }, sdk) => {
        const logger = createEndpointV2Logger();
        const label = devtools.formatOmniPoint(point);
        const transactions = [];
        if (!(config == null ? void 0 : config.readChannelConfigs)) {
          logger.verbose(`readChannelConfigs not defined for ${label}, skipping`);
          return [];
        }
        logger.verbose(`Checking read channels ${label}`);
        for (const { channelId, defaultReadLibrary } of config.readChannelConfigs) {
          const sendAddress = await sdk.getDefaultSendLibrary(channelId);
          logger.verbose(`Checking default send library for channel ${channelId} for ${label}`);
          if (defaultReadLibrary === sendAddress) {
            logger.verbose(
              `Default send library for channel ${channelId} is already set to ${defaultReadLibrary} for ${label}`
            );
            continue;
          } else {
            logger.verbose(
              `Setting default send library for channel ${channelId} to ${defaultReadLibrary} for ${label}`
            );
            transactions.push(await sdk.setDefaultSendLibrary(channelId, defaultReadLibrary));
          }
          const receiveAddress = await sdk.getDefaultReceiveLibrary(channelId);
          logger.verbose(`Checking default receive library for channel ${channelId} for ${label}`);
          if (defaultReadLibrary === receiveAddress) {
            logger.verbose(
              `Default receive library for channel ${channelId} is already set to ${defaultReadLibrary} for ${label}`
            );
            continue;
          } else {
            logger.verbose(
              `Setting default receive library for channel ${channelId} to ${defaultReadLibrary} for ${label}`
            );
            transactions.push(await sdk.setDefaultReceiveLibrary(channelId, defaultReadLibrary));
          }
        }
        return transactions;
      },
      {
        onStart: (logger, [{ point }]) => logger.verbose(`Checking Endpoint default read libraries for ${devtools.formatOmniPoint(point)}`),
        onSuccess: (logger, [{ point }]) => logger.verbose(
          `${ioDevtools.printBoolean(true)} Checked Endpoint default read libraries for ${devtools.formatOmniPoint(point)}`
        ),
        onError: (logger, [{ point }], error) => logger.error(
          `Failed to check Endpoint default read libraries for ${devtools.formatOmniPoint(point)}: ${error}`
        )
      }
    )
  )
);
var configureEndpointV2 = withEndpointV2Logger(
  devtools.createConfigureMultiple(
    configureEndpointV2RegisterLibraries,
    configureEndpointV2DefaultReceiveLibraries,
    configureEndpointV2DefaultSendLibraries,
    configureEndpointV2DefaultReadLibraries
  ),
  {
    onStart: (logger) => logger.info(`Checking EndpointV2 configuration`),
    onSuccess: (logger) => logger.info(`${ioDevtools.printBoolean(true)} Checked EndpointV2 configuration`),
    onError: (logger, args, error) => logger.error(`Failed to check EndpointV2 configuration: ${error}`)
  }
);
var TimeoutSchema = zod.z.object({
  lib: devtools.AddressSchema,
  expiry: devtools.UIntBigIntSchema
});
var configureExecutor = async (graph, createSdk) => devtools.flattenTransactions([await configureExecutorDstConfig(graph, createSdk)]);
var configureExecutorDstConfig = async (graph, createSdk) => devtools.flattenTransactions(
  await Promise.all(
    graph.connections.map(async ({ vector: { from, to }, config }) => {
      const sdk = await createSdk(from);
      const dstConfig = await sdk.getDstConfig(to.eid);
      if (devtools.isDeepEqual(dstConfig, config.dstConfig)) {
        return [];
      }
      return [await sdk.setDstConfig(to.eid, config.dstConfig)];
    })
  )
);
var ExecutorDstConfigPre2_1_27Schema = zod.z.object({
  baseGas: devtools.UIntBigIntSchema,
  multiplierBps: devtools.UIntBigIntSchema,
  floorMarginUSD: devtools.UIntBigIntSchema,
  nativeCap: devtools.UIntBigIntSchema
});
var ExecutorDstConfigPost2_1_27Schema = zod.z.object({
  lzComposeBaseGas: devtools.UIntBigIntSchema,
  lzReceiveBaseGas: devtools.UIntBigIntSchema,
  multiplierBps: devtools.UIntBigIntSchema,
  floorMarginUSD: devtools.UIntBigIntSchema,
  nativeCap: devtools.UIntBigIntSchema
});
var ExecutorDstConfigSchema = zod.z.union([
  ExecutorDstConfigPre2_1_27Schema,
  ExecutorDstConfigPost2_1_27Schema
]);
var configurePriceFeed = async (graph, createSdk) => devtools.flattenTransactions([await configurePriceFeedPriceData(graph, createSdk)]);
var configurePriceFeedPriceData = async (graph, createSdk) => devtools.flattenTransactions(
  await Promise.all(
    graph.connections.map(async ({ vector: { from, to }, config }) => {
      const sdk = await createSdk(from);
      const priceData = await sdk.getPrice(to.eid);
      if (devtools.isDeepEqual(priceData, config.priceData)) {
        return [];
      }
      return [await sdk.setPrice(to.eid, config.priceData)];
    })
  )
);
var PriceDataSchema = zod.z.object({
  priceRatio: devtools.UIntBigIntSchema,
  gasPriceInUnit: devtools.UIntBigIntSchema,
  gasPerByte: devtools.UIntBigIntSchema
});
var configureUln302 = async (graph, createSdk) => devtools.flattenTransactions([
  ...await configureUln302DefaultExecutorConfigs(graph, createSdk),
  ...await configureUln302DefaultUlnConfigs(graph, createSdk)
]);
var configureUln302DefaultExecutorConfigs = async (graph, createSdk) => devtools.flattenTransactions(
  await Promise.all(
    graph.contracts.map(async ({ point, config }) => {
      const sdk = await createSdk(point);
      return Promise.all(
        config.defaultExecutorConfigs.map(([eid, config2]) => sdk.setDefaultExecutorConfig(eid, config2))
      );
    })
  )
);
var configureUln302DefaultUlnConfigs = async (graph, createSdk) => devtools.flattenTransactions(
  await Promise.all(
    graph.contracts.map(async ({ point, config }) => {
      const sdk = await createSdk(point);
      return Promise.all(
        config.defaultUlnConfigs.map(([eid, config2]) => sdk.setDefaultUlnConfig(eid, config2))
      );
    })
  )
);
var Uln302ExecutorConfigSchema = zod.z.object({
  executor: devtools.AddressSchema,
  maxMessageSize: devtools.UIntNumberSchema
});
var Uln302UlnConfigSchema = zod.z.object({
  confirmations: devtools.UIntBigIntSchema,
  requiredDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNThreshold: devtools.UIntNumberSchema
});
var Uln302UlnUserConfigSchema = zod.z.object({
  confirmations: devtools.UIntBigIntSchema.optional(),
  requiredDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNs: zod.z.array(devtools.AddressSchema).optional(),
  optionalDVNThreshold: devtools.UIntNumberSchema.optional()
});

// src/uln302/types.ts
var Uln302ConfigType = /* @__PURE__ */ ((Uln302ConfigType2) => {
  Uln302ConfigType2["Send"] = "send";
  Uln302ConfigType2["Receive"] = "receive";
  return Uln302ConfigType2;
})(Uln302ConfigType || {});
var createUlnReadLogger = () => ioDevtools.createModuleLogger("UlnRead");
var withUlnReadLogger = ioDevtools.createWithAsyncLogger(createUlnReadLogger);
var configureUlnReadDefaultUlnConfigs = withUlnReadLogger(
  devtools.createConfigureNodes(
    withUlnReadLogger(async ({ config, point }, sdk) => {
      const logger = createUlnReadLogger();
      const label = devtools.formatOmniPoint(point);
      const omniTransactions = [];
      if (!config.defaultUlnConfigs) {
        logger.verbose(`defaultUln configuration not set for ${label}, skipping`);
        return [];
      }
      for (const [channelId, ulnConfig] of config.defaultUlnConfigs) {
        logger.verbose(`Setting defaultUln configuration for ${label} on channel ${channelId}`);
        const transaction = await sdk.setDefaultUlnConfig(channelId, ulnConfig);
        omniTransactions.push(transaction);
        logger.verbose(`Set defaultUln configuration for ${label} on channel ${channelId}`);
      }
      return omniTransactions;
    })
  )
);
var configureUlnRead = withUlnReadLogger(
  devtools.createConfigureMultiple(configureUlnReadDefaultUlnConfigs),
  {
    onStart: (logger) => logger.info(`Checking UlnRead configuration`),
    onSuccess: (logger) => logger.info(`${ioDevtools.printBoolean(true)} Checked UlnRead configuration`),
    onError: (logger, args, error) => logger.error(`Failed to check UlnRead configuration: ${error}`)
  }
);
var UlnReadUlnConfigSchema = zod.z.object({
  executor: devtools.AddressSchema,
  requiredDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNThreshold: devtools.UIntNumberSchema
});
var UlnReadUlnUserConfigSchema = zod.z.object({
  executor: devtools.AddressSchema.optional(),
  requiredDVNs: zod.z.array(devtools.AddressSchema),
  optionalDVNs: zod.z.array(devtools.AddressSchema).optional(),
  optionalDVNThreshold: devtools.UIntNumberSchema.optional()
});

exports.DVNDstConfigSchema = DVNDstConfigSchema;
exports.ExecutorDstConfigPost2_1_27Schema = ExecutorDstConfigPost2_1_27Schema;
exports.ExecutorDstConfigPre2_1_27Schema = ExecutorDstConfigPre2_1_27Schema;
exports.ExecutorDstConfigSchema = ExecutorDstConfigSchema;
exports.PriceDataSchema = PriceDataSchema;
exports.TimeoutSchema = TimeoutSchema;
exports.Uln302ConfigType = Uln302ConfigType;
exports.Uln302ExecutorConfigSchema = Uln302ExecutorConfigSchema;
exports.Uln302UlnConfigSchema = Uln302UlnConfigSchema;
exports.Uln302UlnUserConfigSchema = Uln302UlnUserConfigSchema;
exports.UlnReadUlnConfigSchema = UlnReadUlnConfigSchema;
exports.UlnReadUlnUserConfigSchema = UlnReadUlnUserConfigSchema;
exports.configureDVN = configureDVN;
exports.configureDVNDstConfig = configureDVNDstConfig;
exports.configureEndpointV2 = configureEndpointV2;
exports.configureEndpointV2DefaultReadLibraries = configureEndpointV2DefaultReadLibraries;
exports.configureEndpointV2DefaultReceiveLibraries = configureEndpointV2DefaultReceiveLibraries;
exports.configureEndpointV2DefaultSendLibraries = configureEndpointV2DefaultSendLibraries;
exports.configureEndpointV2RegisterLibraries = configureEndpointV2RegisterLibraries;
exports.configureExecutor = configureExecutor;
exports.configureExecutorDstConfig = configureExecutorDstConfig;
exports.configurePriceFeed = configurePriceFeed;
exports.configurePriceFeedPriceData = configurePriceFeedPriceData;
exports.configureUln302 = configureUln302;
exports.configureUln302DefaultExecutorConfigs = configureUln302DefaultExecutorConfigs;
exports.configureUln302DefaultUlnConfigs = configureUln302DefaultUlnConfigs;
exports.configureUlnRead = configureUlnRead;
exports.configureUlnReadDefaultUlnConfigs = configureUlnReadDefaultUlnConfigs;
