// @layerzerolabs/io-devtools@0.1.15/dist/swag/index.mjs
import React2, { useState, useEffect } from 'react';
import { render, Box, Text } from 'ink';
export { render } from 'ink';
import Gradient from 'ink-gradient';
import { stdout } from 'process';
import InkTable from 'ink-table';

// src/swag/printer.tsx
var Logo = () => {
  var _a;
  const [columns, setColumns] = useState((_a = stdout.columns) != null ? _a : 80);
  const logo = columns >= 130 ? LOGO_LARGE : LOGO_SMALL;
  useEffect(() => {
    const handleResize = () => {
      var _a2;
      return setColumns((_a2 = stdout.columns) != null ? _a2 : 80);
    };
    stdout.on("resize", handleResize);
    return () => {
      stdout.off("resize", handleResize);
    };
  }, []);
  return /* @__PURE__ */ React2.createElement(Box, { justifyContent: "center", marginBottom: 5, marginTop: 5 }, /* @__PURE__ */ React2.createElement(Gradient, { name: "rainbow", key: logo }, /* @__PURE__ */ React2.createElement(Text, null, logo)));
};
var LOGO_LARGE = `
     **********                                                                                                                  
   **************                                                                                                                
 ******************                                                                                                              
********************                                                                                                             
*********  *********                                                                                                             
*********  *********          ****                                                 ***********                                   
*********  *********          ****                                                 ***********                                   
   ******  *********          ****        *************    ****  ******   *******       ****    ******    *** **   ******        
 ********  *********          ****      *********** ****  **** ********** *******      ****   *********** ****** **********      
*********  ********           ****     ****    ****  ******** ***   ***** ****        ****   ****   ***** ****  ****    ****     
*********  ******             ****     ****    ****   ******* ********    ****      *****    *********    ***   ****    ****     
*********  *********          ********* ***********   ******   ********** ****     ********************** ***    **********  ****
*********  *********          *********  **********    ****     ********  ****     ***********  ********  ***      *******   ****
*********  *********                                 *****                                                                       
********************                                 ****                                                                        
 ******************                                                                                                              
  ****************                                                                                                               
     **********                                                                                                                  

`;
var LOGO_SMALL = `
     **********     
   **************   
 ****************** 
********************
*********  *********
*********  *********
*********  *********
   ******  *********
 ********  *********
*********  ******** 
*********  ******   
*********  *********
*********  *********
*********  *********
********************
 ****************** 
  ****************  
     **********     

`;
var RecordList = ({
  data = [],
  columns
}) => {
  return /* @__PURE__ */ React2.createElement(React2.Fragment, null, data.map((row, index) => /* @__PURE__ */ React2.createElement(Record, { data: row, columns, key: index })));
};
var Record = ({
  data = {},
  columns = ((_a) => (_a = process.stdout.columns) != null ? _a : 80)()
}) => {
  const padding = 1;
  const labels = Object.keys(data);
  const labelLengths = labels.map(({ length }) => length);
  const maxLabelLength = Math.max(1, ...labelLengths);
  const labelColumnWidth = maxLabelLength + 2 * padding;
  const valueColumnWidth = Math.max(10, columns - labelColumnWidth);
  return /* @__PURE__ */ React2.createElement(
    Box,
    {
      flexDirection: "column",
      borderColor: "grey",
      borderStyle: "single",
      paddingX: 1
    },
    Object.entries(data).map(([key, value]) => {
      return /* @__PURE__ */ React2.createElement(Box, { key }, /* @__PURE__ */ React2.createElement(Box, { width: labelColumnWidth }, /* @__PURE__ */ React2.createElement(Text, { bold: true, wrap: "truncate", color: "magenta" }, key)), /* @__PURE__ */ React2.createElement(Box, { width: valueColumnWidth }, value == null ? /* @__PURE__ */ React2.createElement(Text, { wrap: "wrap", color: "gray" }, "-") : /* @__PURE__ */ React2.createElement(Text, { wrap: "wrap" }, String(value))));
    })
  );
};
var Table = ({ data }) => /* @__PURE__ */ React2.createElement(
  InkTable,
  {
    data,
    header: ({ children }) => /* @__PURE__ */ React2.createElement(Text, { bold: true, color: "magenta" }, children)
  }
);

// src/swag/printer.tsx
var printLogo = () => render(/* @__PURE__ */ React2.createElement(Logo, null)).unmount();
var printRecord = (data) => render(/* @__PURE__ */ React2.createElement(Record, { data })).unmount();
var printRecords = (data) => render(/* @__PURE__ */ React2.createElement(RecordList, { data })).unmount();
var printVerticalTable = (data) => render(/* @__PURE__ */ React2.createElement(Table, { data })).unmount();
var ProgressBar = ({
  columns = ((_a) => (_a = process.stdout.columns) != null ? _a : 80)(),
  progress = 0,
  before = "",
  after = ""
}) => {
  const padding = before.length + after.length;
  const space = Math.max(0, columns - padding);
  const clampedProgress = Math.max(0, Math.min(1, progress));
  const numChars = Math.min(Math.floor(space * clampedProgress), space);
  const chars = CHARACTER.repeat(numChars);
  const numEmptyChars = space - numChars;
  const emptyChars = EMPTY_CHARACTER.repeat(numEmptyChars);
  return /* @__PURE__ */ React2.createElement(Box, null, /* @__PURE__ */ React2.createElement(Text, null, before), /* @__PURE__ */ React2.createElement(Box, { alignItems: "center" }, /* @__PURE__ */ React2.createElement(Gradient, { name: "rainbow" }, /* @__PURE__ */ React2.createElement(Text, null, chars)), /* @__PURE__ */ React2.createElement(Text, null, emptyChars)), /* @__PURE__ */ React2.createElement(Text, null, after));
};
var CHARACTER = "\u2588";
var EMPTY_CHARACTER = "\u2591";
var createProgressBar = (props = {}) => /* @__PURE__ */ React2.createElement(ProgressBar, { ...props });

export { createProgressBar, printLogo, printRecord, printRecords, printVerticalTable };
