// eslint-plugin-jest-extended@2.0.3/lib/rules/prefer-to-be-true.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _utils = require("@typescript-eslint/utils");
var _utils2 = require("./utils");
const isTrueLiteral = node => node.type === _utils.AST_NODE_TYPES.Literal && node.value === true;
var _default = exports.default = (0, _utils2.createRule)({
  name: __filename,
  meta: {
    docs: {
      category: 'Stylistic Issues',
      description: 'Suggest using `toBeTrue()`',
      recommended: false
    },
    messages: {
      preferToBeTrue: 'Prefer using `toBeTrue()` to test value is `true`.'
    },
    fixable: 'code',
    type: 'suggestion',
    schema: []
  },
  defaultOptions: [],
  create(context) {
    return {
      CallExpression(node) {
        const jestFnCall = (0, _utils2.parseJestFnCall)(node, context);
        if ((jestFnCall === null || jestFnCall === void 0 ? void 0 : jestFnCall.type) !== 'expect') {
          return;
        }
        if (jestFnCall.args.length === 1 && isTrueLiteral((0, _utils2.getFirstMatcherArg)(jestFnCall)) && _utils2.EqualityMatcher.hasOwnProperty((0, _utils2.getAccessorValue)(jestFnCall.matcher))) {
          context.report({
            node: jestFnCall.matcher,
            messageId: 'preferToBeTrue',
            fix: fixer => [fixer.replaceText(jestFnCall.matcher, 'toBeTrue'), fixer.remove(jestFnCall.args[0])]
          });
        }
      }
    };
  }
});
