// @layerzerolabs/io-devtools@0.1.15/dist/filesystem/index.js
'use strict';

var winston = require('winston');
var zod = require('zod');
var fs = require('fs');

var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});
var LogLevel = /* @__PURE__ */ ((LogLevel2) => {
  LogLevel2["error"] = "error";
  LogLevel2["warn"] = "warn";
  LogLevel2["info"] = "info";
  LogLevel2["http"] = "http";
  LogLevel2["verbose"] = "verbose";
  LogLevel2["debug"] = "debug";
  LogLevel2["silly"] = "silly";
  return LogLevel2;
})(LogLevel || {});
zod.z.nativeEnum(LogLevel);
var DEFAULT_LOG_LEVEL = "info" /* info */;
var createLogger = (level = DEFAULT_LOG_LEVEL, logFormat = winston.format.cli()) => winston.createLogger({
  level,
  format: logFormat,
  transports: [new winston.transports.Console()]
});
var createModuleLogger = (module, level = DEFAULT_LOG_LEVEL) => createLogger(level, winston.format.combine(prefix({ label: module }), winston.format.cli()));
var prefix = winston.format((info, { label }) => ({
  ...info,
  message: `${label ? `[${label}] ` : ""}${info.message}`
}));
var isDirectory = (path) => {
  try {
    return fs.lstatSync(path).isDirectory();
  } catch (e) {
    return false;
  }
};
var isFile = (path) => {
  try {
    return fs.lstatSync(path).isFile();
  } catch (e) {
    return false;
  }
};
var isReadable = (path) => {
  try {
    return fs.accessSync(path, fs.constants.R_OK), true;
  } catch (e) {
    return false;
  }
};
var importDefault = async (path) => {
  const logger = createModuleLogger("filesystem");
  try {
    logger.debug(`Importing default from '${path}' using dynamic import`);
    const result = await import(path);
    if (result != null && "default" in result) {
      return logger.debug(`Found default in '${path}' using dynamic import`), result.default;
    }
    return logger.debug(`Did not find default in '${path}' using dynamic import, returning the whole thing`), result;
  } catch (error) {
    logger.debug(`Failed to import from '${path}' using dynamic import: ${error}`);
  }
  try {
    logger.debug(`Importing default from '${path}' using require`);
    const result = await __require(path);
    if (result != null && result.__esModule) {
      return logger.debug(`Found default in '${path}' using require`), result.default;
    }
    return logger.debug(`Did not find default in '${path}' using require, returning the whole thing`), result;
  } catch (error) {
    logger.debug(`Failed to import from '${path}' using require: ${error}`);
    throw error;
  }
};

exports.importDefault = importDefault;
exports.isDirectory = isDirectory;
exports.isFile = isFile;
exports.isReadable = isReadable;
